File: frontend/src/lib/pexelsService.ts
import { PexelsResponse, PexelsPhoto } from '@/types/pexels';

const PEXELS_API_KEY = process.env.NEXT_PUBLIC_PEXELS_API_KEY;

if (!PEXELS_API_KEY) {
  console.warn('PEXELS_API_KEY is not set in environment variables');
}

/**
 * Fetch photos from Pexels API based on search query
 */
export async function fetchPexelsPhotos(query: string, perPage = 1): Promise<PexelsPhoto[]> {
  if (!PEXELS_API_KEY) {
    // Return empty array if no API key
    return [];
  }

  try {
    const response = await fetch(
      `https://api.pexels.com/v1/search?query=${encodeURIComponent(query)}&per_page=${perPage}`,
      {
        headers: {
          Authorization: PEXELS_API_KEY,
        },
      }
    );

    if (!response.ok) {
      throw new Error(`Pexels API error: ${response.status}`);
    }

    const data: PexelsResponse = await response.json();
    return data.photos;
  } catch (error) {
    console.error('Error fetching from Pexels:', error);
    return [];
  }
}

/**
 * Get a single photo URL for a place
 */
export async function getPlaceImageUrl(placeName: string, destination: string = ''): Promise<string> {
  if (!PEXELS_API_KEY) {
    // Fallback to Unsplash if no Pexels key
    const searchQuery = encodeURIComponent(`${placeName} ${destination} Egypt`.trim());
    return `https://source.unsplash.com/600x400/?${searchQuery}`;
  }

  try {
    const searchQuery = `${placeName} ${destination} Egypt`.trim();
    const photos = await fetchPexelsPhotos(searchQuery, 1);
    
    if (photos.length > 0 && photos[0].src?.medium) {
      return photos[0].src.medium;
    }
    
    // Fallback to Unsplash if no results
    const fallbackQuery = encodeURIComponent(`${placeName} ${destination} Egypt`.trim());
    return `https://source.unsplash.com/600x400/?${fallbackQuery}`;
  } catch (error) {
    console.error('Error getting place image URL:', error);
    // Fallback to Unsplash on error
    const fallbackQuery = encodeURIComponent(`${placeName} ${destination} Egypt`.trim());
    return `https://source.unsplash.com/600x400/?${fallbackQuery}`;
  }
}

File: src/pexels_service.py
import os
import logging
import requests
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)

PEXELS_API_KEY = os.getenv("PEXELS_API_KEY")

# In-memory cache: key = "place_name|destination"
_image_cache: dict = {}


def fetch_pexels_photos(query: str, per_page: int = 1):
    """Fetch photos from Pexels API based on search query."""
    if not PEXELS_API_KEY:
        logger.warning("[Pexels] No PEXELS_API_KEY found in .env - image search disabled")
        return []

    try:
        url = "https://api.pexels.com/v1/search"
        headers = {"Authorization": PEXELS_API_KEY}
        params = {"query": query, "per_page": per_page}

        response = requests.get(url, headers=headers, params=params)

        if response.status_code != 200:
            logger.error(f"[Pexels] API error: {response.status_code}")
            return []

        data = response.json()
        return data.get("photos", [])

    except Exception as e:
        logger.error(f"[Pexels] Error fetching photos: {e}")
        return []


def get_place_image_url(place_name: str, destination: str = "") -> str | None:
    """Get a single photo URL for a place, with in-memory caching."""
    if not PEXELS_API_KEY:
        return None

    cache_key = f"{place_name}|{destination}"
    if cache_key in _image_cache:
        return _image_cache[cache_key]

    try:
        search_query = f"{place_name} {destination} Egypt".strip()
        photos = fetch_pexels_photos(search_query, 1)

        result = None
        if photos and len(photos) > 0 and "src" in photos[0]:
            result = photos[0]["src"].get("medium")

        _image_cache[cache_key] = result
        return result

    except Exception as e:
        logger.error(f"[Pexels] Error getting place image URL: {e}")
        _image_cache[cache_key] = None
        return None